// cohere-ai/serialization/resources/v2/client/requests/V2ChatStreamRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2ChatStreamRequest = void 0;
const core = __importStar(require("../../../../../core"));
const ChatMessages_1 = require("../../types/ChatMessages");
const Tool2_1 = require("../../types/Tool2");
const V2ChatStreamRequestCitationMode_1 = require("../../types/V2ChatStreamRequestCitationMode");
const ResponseFormat2_1 = require("../../../../types/ResponseFormat2");
exports.V2ChatStreamRequest = core.serialization.object({
    model: core.serialization.string(),
    messages: ChatMessages_1.ChatMessages,
    tools: core.serialization.list(Tool2_1.Tool2).optional(),
    citationMode: core.serialization.property("citation_mode", V2ChatStreamRequestCitationMode_1.V2ChatStreamRequestCitationMode.optional()),
    responseFormat: core.serialization.property("response_format", ResponseFormat2_1.ResponseFormat2.optional()),
    maxTokens: core.serialization.property("max_tokens", core.serialization.number().optional()),
    stopSequences: core.serialization.property("stop_sequences", core.serialization.list(core.serialization.string()).optional()),
    temperature: core.serialization.number().optional(),
    seed: core.serialization.number().optional(),
    frequencyPenalty: core.serialization.property("frequency_penalty", core.serialization.number().optional()),
    presencePenalty: core.serialization.property("presence_penalty", core.serialization.number().optional()),
    k: core.serialization.number().optional(),
    p: core.serialization.number().optional(),
    returnPrompt: core.serialization.property("return_prompt", core.serialization.boolean().optional()),
});
